opa.go

```diff
@@ -0,0 +1,171 @@
+// invoking Eval.
+func New(opts ...func(*config) error) (*OPA, error) {
+	var c = config{
+		MemoryMinPages: 2,
+		MemoryMaxPages: 0,
+		PoolSize:       uint32(runtime.NumCPU()),
+		ErrorLogger:    func(error) {},
+	}
+
+	for _, opt := range opts {
+		if err := opt(&c); err != nil {
+			return nil, err
+		}
+	}
+
+	opa := &OPA{
+		pool:     newPool(c.PoolSize, c.MemoryMinPages, c.MemoryMaxPages),
+		policy:   c.Policy,
+		data:     c.Data,
+		logError: c.ErrorLogger,
+	}
+
+	if len(opa.policy) != 0 {
+		if err := opa.pool.SetPolicyData(opa.policy, opa.data); err != nil {
+			return nil, err
+		}
+	}
+
+	return opa, nil
+}
+
+// SetData updates the data for the subsequent Eval calls.  Returns
+// either ErrNotReady, ErrInvalidPolicyOrData, or ErrInternal if an
+// error occurs.
+func (o *OPA) SetData(v interface{}) error {
+	raw, err := json.Marshal(v)
+	if err != nil {
+		return fmt.Errorf("%v: %w", err, ErrInvalidPolicyOrData)
+	}
+
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	return o.setPolicyData(o.policy, raw)
+}
+
+// SetPolicy updates the policy for the subsequent Eval calls.
+// Returns either ErrNotReady, ErrInvalidPolicy or ErrInternal if an
+// error occurs.
+func (o *OPA) SetPolicy(p []byte) error {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	return o.setPolicyData(p, o.data)
+}
+
+// SetPolicyData updates both the policy and data for the subsequent
+// Eval calls.  Returns either ErrNotReady, ErrInvalidPolicyOrData, or
+// ErrInternal if an error occurs.
+func (o *OPA) SetPolicyData(policy []byte, data *interface{}) error {
+	var raw []byte
+	if data != nil {
+		var err error
+		raw, err = json.Marshal(*data)
+		if err != nil {
+			return fmt.Errorf("%v: %w", err, ErrInvalidPolicyOrData)
+		}
+	}
+
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	return o.setPolicyData(policy, raw)
+}
+
+func (o *OPA) setPolicyData(policy []byte, data []byte) error {
+	if err := o.pool.SetPolicyData(policy, data); err != nil {
+		return err
+	}
+
+	o.policy = policy
+	o.data = data
+	return nil
+}
+
+// Eval evaluates the policy with the given input, returning the
+// evaluation results. If no policy was configured at construction
+// time nor set after, the function returns ErrNotReady.  It returns
+// ErrInternal if any other error occurs.
+func (o *OPA) Eval(ctx context.Context, input *interface{}) (interface{}, error) {
+	instance, err := o.pool.Acquire(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	defer o.pool.Release(instance)
+
+	result, err := instance.Eval(ctx, input)
+	if err != nil {
+		return nil, fmt.Errorf("%v: %w", err, ErrInternal)
+	}
+
+	return result, nil
+}
+
+// EvalBool evaluates the boolean policy with the given input. The
+// possible error values returned are as with Eval with addition of
+// ErrUndefined indicating an undefined policy decision and
+// ErrNonBoolean indicating a non-boolean policy decision.
+func (o *OPA) EvalBool(ctx context.Context, input *interface{}) (bool, error) {
```

Comment:
I could imagine this flavor of APIs proliferating. At the same time it doesn't do much on top of the low-level `Eval` API. I wonder if splitting this into a package-level function would be a better separation of duty.

Reply:
Done.